text/0000-mark-cf.md

```diff
@@ -0,0 +1,159 @@
+
+As the mark CF satisfies the need of keeping transaction status, these `Lock` and `Rollback` records in the write CF are only for conflict checks. Therefore, once the records are not the latest version, we can safely remove them from the write CF.
+
+**NB.** There are a few cases when the latest version is not enough for conflict checks. For example, `Rollback` records may be skipped when checking newer versions when prewriting non-pessimistic keys in pessimistic transactions. In this case, we also need to read the mark CF to confirm.
+
+#### Checking the write CF early
+
+Write CF records are typically written in the commit command, during which we don't read the write CF before. So, it will bring extra cost if we check whether the latest version can be removed in the commit command.
+
+We can move the check forward to the conflict check process in acquiring pessimistic locks or prewriting. During the conflict check, we can see whether the latest record is a `Lock` or `Rollback` record. If so, we record its commit TS in the lock:
+
+```rust
+pub struct Lock {
+    ...
+    /// Commit TS of a `Lock` or `Rollback` record which is the latest version of the key
+    pub latest_mark_ts: TimeStamp,
+}
+```
+
+Then, when the lock is finally committed into a write record, we can delete the stale `Lock` or `Rollback` record without additionally reading the write CF again.
+
+#### Overwriting the latest `Lock` or `Rollback` (to be discussed)
+
+This is a tricky optimization and I really doubt if it is really worth.
+
+If a new `Lock` record is written to the write CF while the latest version is also a `Lock` or `Rollback`, instead of removing the previous version, just overwrite that record and add a `real_commit_ts` to the record. When checking write conflicts, we should parse the value and check the real commit TS because the timestamp encoded into the key may be not accurate.
+
+It may help reduce tombstones but breaks too many assumptions before.
+
+### Reading the mark CF
+
+Putting `Lock` and `Rollback` records in the mark CF does not introduce extra seek operations to the happy path of transactions. But it does when we talk about resolving locks.
+
+Now, the records that represent transaction status spread in both the write CF and the mark CF.
+
+In `CheckTxnStatus`, we need to read both CFs of the primary key for the status of the transaction.
+
+In `CheckSecondaryKeys`, we need to check both CFs of all the given secondary keys to know whether some keys are already committed or rolled back.
+
+And when prewrite raises an error or we are prewriting non-pessimistic keys in a retry, we also need the precise status of the key to guarantee idempotence. This also requires to read both the write and mark CFs.
+
+Luckily, all of these don't happen frequently in production. The extra cost is not a big issue.
```

Comment:
As the key format is `{user_key}{start_ts}` and it's different from keys is write cf, maybe we could describe a bit more details about the conflict check process here.

Reply:
Conflict checking itself only needs to read the write CF for the maximum version. When we need to read the mark CF, we are confirming whether the transaction we are prewriting has been rolled back. So, it's always a point get according to the `start_ts` on the mark CF.